### PR TITLE
Implementing Cache

### LINKED ISSUE #9: Data Cache for Better Data I/O
Currently, data is sliced from a file _every_ integration period. That is, if the integration time is 60 seconds, then `RadClass` will slice 60 rows at every step.

This could be alleviated by creating a data cache. If `cache_size = 1000`, then 1000 rows will be sliced, and the 60 rows for integration will be indexed from this, greatly reducing the number of data i/o actions.

The challenge is keeping track of indexing in the cache. When data is collected (either by cache or slicing) through `DataSet.data_slice()`, the relative indexes of the `.h5` file are not maintained. For example, if I use `DataSet.data_slice()` to collect rows 1000 to 1060, it will gather 60 rows. However, the indices will no longer be 1000-1060 but 0-60. This is not a problem when the cache is the same size as the integration (the current implementation) but I need a _new_ method of keeping track of indices in the index-reset cache. Open to suggestion!

### PR DESCRIPTION
**Pull request by stompsjo**
This PR includes additions to `RadClass` that implement a `cache_size` to index more data than required by `integration`. This improves performance by reducing the number of indexes pulled from data files. This includes a new `RadClass` method called `run_cache` that updates the rows stored in memory from the data file and keeps track of the new start position/timestamp. If the requisite rows are not found in cache memory, the cache is reinitialized.

A test of implementation is included in `test_RadClass.py`, which simply re-runs `test_integration` with a `cache_size` and double-checks that the answer is the same.

### CODE REVIEW

File: RadClass/RadClass.py [hunk 1]
```diff
@@ -151,6 +160,17 @@ def march(self):
 
         return running
 
+    def run_cache(self):
+        start_i, = np.where(self.processor.timestamps == self.working_time)
+        start_i = start_i[0]
+        if start_i + self.cache_size >= len(self.processor.timestamps):
+            end_i = len(self.processor.timestamps) - 1
+        else:
+            end_i = start_i + self.cache_size
```
Comment (gonuke) [inline RadClass/RadClass.py]: Use `min()`

File: RadClass/RadClass.py [hunk 2]
```diff
@@ -90,13 +90,22 @@ def collapse_data(self, rows):
...
-        data_matrix = self.processor.data_slice(self.datapath, rows)
-
+        data_matrix = np.empty((len(rows), len(self.cache[0])))
         # normalize by live times to produce count rate data
         # processor.live can be indexed by appropriate timestamps but
         # data_matrix only has indices for rows.
-        data_matrix = data_matrix / self.processor.live[rows][:, None]
-
+        for i, row in enumerate(rows):
+            if row not in self.cache_rows:
```
Comment (gonuke) [inline RadClass/RadClass.py]: Can't you check that your cache contains what you need once before running the loop?

File: RadClass/RadClass.py [hunk 3]
```diff
@@ -90,13 +90,22 @@ def collapse_data(self, rows):
...
+                self.run_cache()
+            idx = np.where(self.cache_rows == row)[0][0]
```
Comment (gonuke) [inline RadClass/RadClass.py]: This is expensive given that after you reset the cache, this is just incrementing by 1 each time through the loop

**Review by gonuke — CHANGES_REQUESTED** [review]
I might need to get a better understanding of what you are trying to accomplish, conceptually, and then we can talk about some standard patterns for this kind of thing

**Comment by stompsjo** [comment]
@gonuke The pushed changes address your feedback. Row indices (in the form of `rows_idx`) is still passed to `collapse_data` but now the cache is re-initialized, if required, outside the method. `self.cache_idx` keeps track of the indices in cache, which can be substracted from `rows_idx` (as an offset) to properly index rows for integration. A for-loop is no longer necessary.

File: RadClass/RadClass.py [hunk 4]
```diff
@@ -122,8 +124,14 @@ def collect_rows(self):
                     len(self.processor.timestamps) - 1)
 
         # enumerate number of rows to integrate exclusive of the endpoint
-        rows = np.arange(start_i, end_i)
-        return rows
+        rows_idx = np.arange(start_i, end_i)
+
+        # check if rows are stored in cache
+        if np.count_nonzero(np.isin(self.cache_idx,
```
Comment (gonuke) [inline RadClass/RadClass.py]: Can't we just compare the end of `cache_idx` to `end_i` somehow?  This seems bulky and complicated? Maybe I'm missing something?

File: RadClass/RadClass.py [hunk 5]
```diff
@@ -151,14 +160,24 @@ def march(self):
 
         return running
 
+    def run_cache(self):
+        start_i, = np.where(self.processor.timestamps == self.working_time)
```
Comment (gonuke) [inline RadClass/RadClass.py]: Should we be keeping `start_i` up-to-date as a state variable of `self` rather than having to re-find it periodically?

**Review by gonuke — CHANGES_REQUESTED** [review]
Just a couple of comments/questions.  There may be a case to make a Cache class, but that's not a critical path question.

File: RadClass/RadClass.py (hunk 4 above)
Comment (stompsjo) [reply RadClass/RadClass.py]: This ensures that *every* row (or row index) is stored in cache, but maybe that's too exhaustive and just checking for the end is sufficient. How about this?

```
if end_i not in self.cache_idx:
     self.run_cache()
```

File: RadClass/RadClass.py (hunk 5 above)
Comment (stompsjo) [reply RadClass/RadClass.py]: `start_i` is already kept up-to-date as a state variable, in a way, as `self.working_time`, which is the value equivalent of the index, `start_i`. Similar operations happen in liones 116 and 146. I could collapse all of these operations to one by adding an `np.where()` execution in `march()` when `self.working_time` is updated?

File: RadClass/RadClass.py (hunk 4 above)
Comment (gonuke) [reply RadClass/RadClass.py]: I agree that your original case is more robust, but I think you do everything linearly, so this proposal should be enough

File: RadClass/RadClass.py (hunk 5 above)
Comment (gonuke) [reply RadClass/RadClass.py]: Or increment `start_i` by some amount each time you use it?

**Comment by stompsjo** [comment]
@gonuke Both comments are addressed/implemented in the newest commit.

**Review by gonuke — APPROVED** [review]
Thanks @stompsjo

**Pull request merged**